### DISCUSSION TITLE
Special MQTT headers in MQTT 5

### DISCUSSION THREAD

**Discussion by dimabarbul** [eclipse-ditto/ditto#1757] [Q&A] [answered]
I'm not sure if this is expected behavior or not, so don't want to create issue.

I have question about special MQTT headers: mqtt.topic, mqtt.qos, mqtt.retain.

1) When I send message and set, for example, mqtt.retain header (using headerMapping or payload mapper), it will be applied (i.e., the message will be retained) **and added to MQTT user properties**. I'd expect these special headers to not be added to user properties. I try to set up communication with devices which contains zero knowledge about Ditto stuff, so to my use case it would be good to not have any extra user properties.

2) When I receive message using MQTT 5, the headers are set accordingly to message features, for example, mqtt.topic will contain topic name that the message was received on. But if the message has user property mqtt.topic, it will override special meaning of the header and contain whatever user property contains.

So my question is: are the 2 points mentioned above expected or is it something that should/could be corrected?

**Comment by thjaeckle** [comment, marked as answer]
>  I'd expect these special headers to not be added to user properties

Yes, I agree - that should not be the case.
For other connection types, [like AMQP 1.0 we explicitly do not add such "protocol headers" to the "user headers"](https://github.com/eclipse-ditto/ditto/blob/b80497dfc6cc87839d19143f0b9ddc995edfea4f/connectivity/service/src/main/java/org/eclipse/ditto/connectivity/service/messaging/amqp/JMSPropertyMapper.java#L77).
Seems that for MQTT5 this was forgotten.

The second one seems also not wanted, although it is quite a corner case ..
